src/CanvasPattern.cc

```diff
@@ -4,122 +4,115 @@
 
 #include "Canvas.h"
 #include "Image.h"
+#include "InstanceData.h"
 
-using namespace v8;
+using namespace Napi;
 
 const cairo_user_data_key_t *pattern_repeat_key;
 
-Nan::Persistent<FunctionTemplate> Pattern::constructor;
-Nan::Persistent<Function> Pattern::_DOMMatrix;
-
 /*
  * Initialize CanvasPattern.
  */
 
 void
-Pattern::Initialize(Nan::ADDON_REGISTER_FUNCTION_ARGS_TYPE target) {
-  Nan::HandleScope scope;
+Pattern::Initialize(Napi::Env& env, Napi::Object& exports) {
+  Napi::HandleScope scope(env);
+  InstanceData* data = env.GetInstanceData<InstanceData>();
 
   // Constructor
-  Local<FunctionTemplate> ctor = Nan::New<FunctionTemplate>(Pattern::New);
-  constructor.Reset(ctor);
-  ctor->InstanceTemplate()->SetInternalFieldCount(1);
-  ctor->SetClassName(Nan::New("CanvasPattern").ToLocalChecked());
-  Nan::SetPrototypeMethod(ctor, "setTransform", SetTransform);
+  Napi::Function ctor = DefineClass(env, "CanvasPattern", {
+    InstanceMethod<&Pattern::setTransform>("setTransform")
+  });
 
   // Prototype
-  Local<Context> ctx = Nan::GetCurrentContext();
-  Nan::Set(target, Nan::New("CanvasPattern").ToLocalChecked(), ctor->GetFunction(ctx).ToLocalChecked());
-  Nan::Set(target, Nan::New("CanvasPatternInit").ToLocalChecked(), Nan::New<Function>(SaveExternalModules));
-}
-
-/*
- * Save some external modules as private references.
- */
-
-NAN_METHOD(Pattern::SaveExternalModules) {
-  _DOMMatrix.Reset(Nan::To<Function>(info[0]).ToLocalChecked());
+  exports.Set("CanvasPattern", ctor);
+  data->CanvasPatternCtor = Napi::Persistent(ctor);
 }
 
 /*
  * Initialize a new CanvasPattern.
  */
 
-NAN_METHOD(Pattern::New) {
-  if (!info.IsConstructCall()) {
-    return Nan::ThrowTypeError("Class constructors cannot be invoked without 'new'");
+Pattern::Pattern(const Napi::CallbackInfo& info) : ObjectWrap<Pattern>(info), env(info.Env()) {
```

Comment:
A crash happens with `new canvas.CanvasPattern()` or `ctx.createPattern()`.

Reply:
I'm not able to reproduce any crash. I get `Uncaught TypeError: Image or Canvas expected` for both of those, which is correct. I also tried passing a real image, and verified that is painting correctly.

Reply:
```
> canvas = createCanvas(1,1)
[Canvas 1x1]
> canvas.getContext("2d").createPattern()
# Exception thrown at 0x0000000068DF09A6 (libcairo-2.dll) in node.exe: 0xC0000005: Access violation reading location 0xFFFFFFFFFFFFFFFF
```

Turns out a crash happens with the same code in master also, but with a different cause. Can deal with later.